rfcs/0007-synchronous-commit.md

```diff
@@ -0,0 +1,442 @@
+    WAL->>Flusher: Flush WAL
+    Flusher-->>WAL: ACK FlushWAL message
+    WAL->>MemTable: Merge WAL to MemTable
+```
+
+In this proposal, "Commit to make it visible" effectively means "applying the changes to the MemTable", whenever the change is applied to MemTable, it's considered as committed.
+
+The changes to the read path are minimal. Readers with `ReadWatermark::LastCommitting` will continue to read data from the WAL before accessing the MemTable. For reads with `ReadWatermark::LastCommitted`, `ReadWatermark::LastLocalPersisted`, or `ReadWatermark::LastRemotePersisted`, readers will first check the MemTable, using different sequence numbers as watermarks to determine visibility.
+
+On the write side, after the data is appended to the WAL, it'll be applied to MemTable as soon as possible to make the change to be visible to readers.
+
+The sequence diagram is like this:
+
+```mermaid
+sequenceDiagram
+    participant WriteBatch
+    participant WAL
+    participant MemTable
+    participant Flusher
+
+    WriteBatch->>WAL: append write op to current WAL
+    Note over WAL: become visible to readers with ReadWatermark::LastCommitting
+
+    WAL->>WAL: Maybe trigger a flush
+    Note over WAL: when it reaches flush.interval or flush.size
+
+    WAL->>Flusher: Flush WAL
+
+
+    Flusher-->>WAL: ack FlushWAL message, increment last wal flushed position
+
+    WAL->>MemTable: apply write op to MemTable
+    Note over MemTable: become visible to readers with ReadWaterMark::LastCommitted
+
+    alt sync is Remote
+        WriteBatch->>WAL: await last applied position
+    end
+
+    MemTable-->>WriteBatch: wake up waiters of last applied position
+```
+
+### Handling No WAL Write
+
+It's not possible to have Sync Commit when WAL is disabled. So let's not allow users to enable `sync` when WAL is disabled.
+
+When `sync` is not possible to set, all the writes will be considered as Committed immediately, and become visible to readers in the MemTable.
+
+Users can still use `await_durable` to wait for the write to be durably persisted on the writer side.
+
+### Handling WAL Write Failures
+
+It's more likely to have WAL write failures on object storage like S3 than local disk.
+
+We do not have to fail the write operation immediately when the WAL write fails. Instead, we can retry the write operation with a backoff.
+
+However, it's possible to have some failure on flushing WAL to object storage which last for several minutes or even longer, and without a local disk which might help to buffer the WAL during the failure.
+
+In this case, we can still buffer the WAL in memory as long as possible. When the failure is resolved, we can resume the flush operation.
+
+However, the memory is limited, and we can't buffer the WAL in memory forever. We already have a threshold setting `max_unflushed_bytes` for this. When this threshold is reached, we can't buffer the WAL anymore, and we have to mark the db into an read-only state.
```

Comment:
Rather than marking the DB as read-only, I think we should apply backpressure for writers by either throwing an exception or just blocking the write.